lib/spack/spack/spec.py

```diff
@@ -3871,6 +3871,13 @@ def _cmp_iter(self):
         # TODO: spec hashing.
         yield self.process_hash() if self.concrete else None
 
+        # If there is ever a breaking change to hash computation, whether accidental or purposeful,
+        # two specs can be identical modulo DAG hash, depending on what time they were concretized
+        # From the perspective of many operation in Spack (database, build cache, etc) a different
+        # DAG hash means a different spec. Here we ensure that two otherwise identical specs, one
+        # serialized before the hash change and one after, are considered different.
+        yield self.dag_hash() if self.concrete else None
```

Comment:
Can we speed up spec comparison by yielding this *before* `_cmp_node`?  process_hash isn’t necessarily cached, but this always is if present, so it should cost nothing to put first.

Reply:
then {x, y} == {x} when x includes test deps but y does not

Reply:
Created an issue here https://github.com/spack/spack/issues/48893

Reply:
We should split `<` and `==` to make that work. I've moved it above `process_hash` so at least that computation is avoided for concrete specs that differ on dag hash.

Reply:
In practice the slowdown is the same (4%)